llama.cpp

```diff
@@ -2731,6 +2817,44 @@ struct llama_context * llama_new_context_with_model(
         }
     }
 
+#ifdef GGML_USE_METAL
+    if (params.n_gpu_layers > 0) {
+        void * data_ptr  = NULL;
+        size_t data_size = 0;
+
+        if (params.use_mmap) {
+            data_ptr  = ctx->model.mapping->addr;
+            data_size = ctx->model.mapping->size;
+        } else {
+            data_ptr  = ggml_get_mem_buffer(ctx->model.ctx_metal);
+            data_size = ggml_get_mem_size  (ctx->model.ctx_metal);
+        }
+
+        const size_t max_size = ggml_get_max_tensor_size(ctx->model.ctx_metal);
+
+        printf("%s: max tensor size = %8.2f MB\n", __func__, max_size/1024.0/1024.0);
+
+#define LLAMA_METAL_CHECK_BUF(result)                                          \
+        if (!(result)) {                                                           \
+            fprintf(stderr, "%s: failed to add buffer\n", __func__);               \
+            llama_free(ctx);                                                       \
+            return NULL;                                                           \
+        }
+
+        LLAMA_METAL_CHECK_BUF(ggml_backend_metal_map_buffer(ctx->model.backend_metal, "data", data_ptr, data_size, max_size));
+
+        struct ggml_backend_buffer * buf_compute = ctx->buf_compute_metal->backend_buffer;
+        struct ggml_backend_buffer * buf_kv      = ctx->kv_self.buf->backend_buffer;
+
+        LLAMA_METAL_CHECK_BUF(ggml_backend_metal_map_buffer(ctx->model.backend_metal, "eval", buf_compute->backend_data, buf_compute->backend_size, 0));
+        LLAMA_METAL_CHECK_BUF(ggml_backend_metal_map_buffer(ctx->model.backend_metal, "kv",   buf_kv->backend_data,      buf_kv->backend_size,      0));
```

Comment:
The mapping is not needed only for `mmap`. The Apple Silicon chips have unified memory. This means that the same memory block can be read and write both by the CPU and the GPU. So even if we don't use `mmap`, we want the CPU allocated buffers to be used by Metal directly.

Reply:
I understand that's the way the Metal backend works currently, but the backends interface is not designed to work in this way. You are going to find a lot of problems if you try to implement it like this. If the Metal backend is capable of doing matrix-matrix multiplication (even if it is slow), I suggest that you implement it in the simplest way possible for now: implement `alloc_buffer`, `get_tensor` and `set_tensor`, and use the `data` member of the tensors in the Metal kernels directly instead of mapping addresses. The ggml-cuda backend should be a good example of this, just replace the CUDA memory allocation and copy functions with the Metal ones.
After that works, we can add the changes necessary to support Accelerate again.

Reply:
Ok, let me try to do it like this. Good thing is I just got it running and thanks to the splits I can now run part of the model on the CPU and the rest on the GPU which was not possible before.

Let's see if I can avoid all the hacks as you suggest.